app/models/key.rb

```diff
@@ -83,15 +83,39 @@ def generate_fingerprint
 
     cmd_status = 0
     cmd_output = ''
+    explicit_fingerprint_algorithm = false
     Tempfile.open('gitlab_key_file') do |file|
       file.puts key
       file.rewind
-      cmd_output, cmd_status = popen(%W(ssh-keygen -lf #{file.path}), '/tmp')
+
+      # OpenSSH 6.8 introduces a new default output format for fingerprints.
+      # Check the version and decide which command to use.
+      version_output, version_status = popen(%W(ssh -V))
+      if version_status.zero?
+        out, _ = version_output.scan /.*?(\d)\.(\d).*?,/
+        major, minor = out[0], out[1]
+        if major.to_i > 6 or (major.to_i == 6 and minor.to_i >= 8)
+          explicit_fingerprint_algorithm = true
+        end
+      end
+
+      if explicit_fingerprint_algorithm
+        cmd_output, cmd_status = popen(%W(ssh-keygen -E md5 -lf #{file.path}), '/tmp')
+      else
+        cmd_output, cmd_status = popen(%W(ssh-keygen -lf #{file.path}), '/tmp')
+      end
     end
 
     if cmd_status.zero?
-      cmd_output.gsub /(\h{2}:)+\h{2}/ do |match|
-        self.fingerprint = match
```

Comment:
On the other hand, anchoring against the MD5 prefix seems like a nice defensive idea.